r/R/dplyr.R

```diff
@@ -126,7 +124,18 @@ filter.arrow_dplyr_query <- function(.data, ..., .preserve = FALSE) {
   filters <- lapply(filts, function (f) {
     # This should yield an Expression as long as the filter function(s) are
     # implemented in Arrow.
-    try(eval_tidy(f, dm), silent = TRUE)
+    tryCatch(eval_tidy(f, dm), error = function(e) {
+      # Look for the cases where bad input was given, i.e. this would fail
+      # in regular dplyr anyway
+      msg <- conditionMessage(e)
+      if (grepl("object '.*'.not.found", msg)) {
```

Comment:
Why do you need this?

Reply:
There are two reasons a filter could have failed at this stage: (1) operations that are supported in R and would work in dplyr on a data.frame in R but that aren't (yet) supported in Arrow; and (2) user error, stuff that is invalid even if the data were already in R. These are (at least two of) the ways that the latter case manifests. Previously, we were intercepting variable references that don't exist in the Dataset and erroring earlier, but to allow the scoping to find the other variable you defined in the global environment, I removed that check, let the expression try to evaluate, and catch the exception.

In Python I'd write this something like:

```python
try:
    eval_tidy(f, dm)
except NotFound:
    # User error, so tell them
    raise
except:
    raise ArrowError("Filter expression %s not supported in Arrow" % str(f))
```

R doesn't (generally) have exception classes, so sniffing the exception message is necessary.

Reply:
In that case, it seems advisable to have a GUID here to make the sniffing easier and collisions less likely

Reply:
I don't follow. These aren't exceptions that my code is raising, so I can't choose how they raise or with what messages.